base/acme/src/main/java/org/dogtagpki/acme/server/ACMEEngineConfigFileSource.java

```diff
@@ -96,41 +98,91 @@ public void run() {
+                // changed, and that we were unable to read it, but after the sleep
+                // it might be OK.  So let's try and read it, but ignore failures.
+                try {
+                    loadFile();
+                } catch (Throwable e) {
+                    // ignore
+                }
+
+            }
+
+            try (WatchService watchService = FileSystems.getDefault().newWatchService()) {
+                // You can only watch a directory, not a file.  So we must
+                // watch the parent dir and check the event is for the file.
+                path.getParent().register(watchService, StandardWatchEventKinds.ENTRY_MODIFY);
+
+                // so far so good; reset the delay
+                logger.info("ACMEEngineConfigSource: watching " + filename);
+                delay_ms = 0;  // reset the delay
+
+                while (true) {
+                    logger.debug("ACMEEngineConfigSource: something happened");
+                    WatchKey wk = watchService.take();
+                    for (WatchEvent<?> event : wk.pollEvents()) {
+                        if (event.kind() == StandardWatchEventKinds.ENTRY_MODIFY) {
+                            if (((Path) event.context()).endsWith(path.getFileName())) {
+                                logger.debug("ACMEEngineConfigSource: file modified");
+                                // sleep for a moment before reading file; this seems
+                                // to avoid a race situation observed with vi/vim.
+                                Thread.sleep(1000);
+                                loadFile();
+                            }
                         }
                     }
-                }
 
-                if (!wk.reset()) {
-                    // watch key could not be reset; break loop
-                    break;
+                    if (!wk.reset()) {
+                        // watch key could not be reset; break inner loop
+                        logger.warn("ACMEEngineConfigSource: watch key could not be reset");
+                        break;
+                    }
                 }
+            } catch (InterruptedException e) {
+                // probably due to shutdown(), in which case monitor == false
+                // and the outer while loop will terminate
+                logger.info("ACMEEngineConfigSource: file monitoring interrupted");
+            } catch (Throwable e) {
+                logger.error(
+                    "ACMEEngineConfigFileSource: caught exception while monitoring file",
+                    e
+                );
+            }
+
+            // Something went wrong, or we were interrupted by shutdown().
+            // Set or increase the delay, then loop back and either try
+            // again, or in case of shutdown() the loop condition fails.
+            if (delay_ms == 0) {
+                delay_ms = 1000;
+            } else {
+                delay_ms *= 2;
```

Comment:
I'd suggest adding an upper bound so the delay doesn't become too big that it practically stops monitoring.

Reply:
Yep, I'll clamp it at 5 mins.